File: python/112-Path Sum.py
from bintree_helper import *

class Solution:
    # def hasPathSum(self, root, sum: int) -> bool:
    #     "DFS, Time: O(n), Space: O(n)"
    #     if root is None: return False
    #     result = False
    #     def dfs(node, total):
    #         nonlocal result
    #         if result:
    #             return
    #         total += node.val
    #         if node.left is None and node.right is None:
    #             if total == sum: result = True
    #         if node.left: dfs(node.left, total)
    #         if node.right: dfs(node.right, total)
    #     dfs(root, 0)
    #     return result
    def hasPathSum(self, root, sum: int) -> bool:
        "DFS, Time: O(n), Space: O(n)"
        if root is None:
            return False
        if root.left is None and root.right is None:
            return sum == root.val
        return self.hasPathSum(root.left, sum-root.val) or self.hasPathSum(root.right, sum-root.val)

root = deserialize("[5,4,8,11,null,13,4,7,2,null,null,null,1]")
result = Solution().hasPathSum(root, 22)
assert result == True

File: cpp/112. Path Sum.cpp
#include <bits/stdc++.h>
using namespace std;
typedef long long ll;
typedef vector<int> vi;
typedef vector<vector<int>> vvi;
typedef pair<int, int> pi;


struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode() : val(0), left(nullptr), right(nullptr) {}
    TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
    TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
};

class Solution {
public:
    bool hasPathSum(TreeNode* root, int sum) {
        if(!root) return false;
        int remain = sum - root->val;   
        if(!root->left && !root->right) {
            if(remain == 0) return true;
            return false;
        }
        if(hasPathSum(root->left, remain)) return true;
        if(hasPathSum(root->right, remain)) return true;
        return false;
    }
};

int main() {
    ios::sync_with_stdio(0); 
    cin.tie(0);

    
}